### PR TITLE
Unify simulate/doFill/doDrain parameters with ActionType

### PR DESCRIPTION
**Pull request by mezz**
Using booleans for the `simulate`, `doFill`, and `doDrain` argument can be a bit confusing to read:
```java
fluidHandler.drain(Fluid.BUCKET_VOLUME, true);
```
Does it mean "true" to simulate the drain, or "true" to do the drain?
Adding to the issue, the boolean for item transactions and fluid transactions are the opposite of each other.

This PR unifies all of these by using an enum instead of a boolean, which is much harder to mess up:
```java
fluidHandler.drain(Fluid.BUCKET_VOLUME, ActionType.EXECUTE);
```

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/EnumSimulate.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common;
+
+public enum EnumSimulate
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/EnumSimulate.java]: I thought we were trying to move away from smurf naming. How about `ActionSimulation`?

**Review by diesieben07 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/common/EnumSimulate.java (hunk 1 above)
Comment (bs2609) [reply src/main/java/net/minecraftforge/common/EnumSimulate.java]: `OperationType`/`ActionType`?

Comment (mezz) [reply src/main/java/net/minecraftforge/common/EnumSimulate.java]: I'm definitely open to better names, I had a hard time naming this and it's still not very good.
We also have some time before 1.13 so might as well pick a good one.

Comment (liach) [reply src/main/java/net/minecraftforge/common/EnumSimulate.java]: I suggest spliting these methods into twos instead.

Comment (mezz) [reply src/main/java/net/minecraftforge/common/EnumSimulate.java]: I think `ActionType` sounds good, it's pretty short too.
Splitting methods into two isn't a great solution. It's best to share the code internally for simulate and execute, so splitting the public methods in two would just mean having the same code as this PR but with extra convenience methods publicly.

**Comment by mezz** [comment]
Changed everything to `ActionType` instead of `EnumSimulate`

**Comment by LemmaEOF** [comment]
Oh, while you're at it, Mezz, would it be possible to clean all the nulls out of fluids and add the `isEmpty()` stuff that ItemStacks have?

**Comment by liach** [comment]
I suggest immutable fluid stacks for those `isEmpty` things.

**Comment by HenryLoenwind** [comment]
Real immutable? So with no setter for the amount?

Possible, but then I'd really want a wrapper object `FluidStackHolder { get/setFluidStack() }...

Oh, and FluidStack.getFluid() would really work much better if it was truly @Nonnull (with no "fluid.AIR" placeholder).

**Comment by LemmaEOF** [comment]
Since it looks like 1.12.2 is gonna be an LTS, I think it'd be good to take the long time to 1.13 to do a total fluid rewrite/fix. Immutable fluids, wrappers, fixing doFill/Drain, unification fixing canonicals, #4678 and other helpful additions, and just stuff like that.

**Comment by mezz** [comment]
Let's keep big ideas like "rewriting the fluids" somewhere else please, like a new issue. That's a whole other can of worms :D

**Comment by liach** [comment]
Hmm, can we add stale exclusion for certain version tags in the future?

**Comment by mezz** [comment]
I'm just assigning 1.13 stuff to myself for now.

**Comment by KingLemming** [comment]
Conceptually I like this - I guess my only possible qualm is the naming of the method which determines if the enum is execute - hasSideEffects().

Offhand that sounds to me like there are *additional* effects beyond just execution. I kind of think action.isExecute() works well enough for the simple action == EXECUTE logic.

**Comment by tterrag1098** [comment]
Please re-target at 1.13,x branch.

**Comment by mezz** [comment]
I started a branch of this for 1.13 a while ago, but it can't really be done without also fully implementing all the other fluid stuff so I got stuck.

**Comment by LexManos** [comment]
This has been addressed in the Fluid re-write.

**Pull request closed without merging**